java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -0,0 +1,188 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.symexecengine;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ExecutionState {
+
+  private final IssuableSubscriptionVisitor check;
+  ExecutionState parent;
+  private SetMultimap<Symbol, Value> reachableValues = HashMultimap.create();
+  private SetMultimap<Symbol, Value> unreachableValues = HashMultimap.create();
+  /**
+   * List of symbol that were declared within this execution state.
+   */
+  private List<Symbol> definedInState = Lists.newArrayList();
+  private Map<Value, State> stateOfValue = Maps.newHashMap();
+
+  public ExecutionState(ExecutionState executionState) {
+    this.parent = executionState;
+    this.check = executionState.check;
+    this.reachableValues = HashMultimap.create(executionState.reachableValues);
+    this.unreachableValues = HashMultimap.create(executionState.unreachableValues);
+  }
+
+  public ExecutionState(IssuableSubscriptionVisitor check) {
+    this.check = check;
+  }
+
+  public void defineSymbol(Symbol symbol) {
+    definedInState.add(symbol);
+  }
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Symbol symbol : executionState.reachableValues.keys()) {
+      this.reachableValues.putAll(symbol, executionState.reachableValues.get(symbol));
+    }
+    for (Symbol symbol : executionState.unreachableValues.keys()) {
+      this.unreachableValues.putAll(symbol, executionState.unreachableValues.get(symbol));
+    }
+
+    for (Symbol symbol : unreachableValues.keys()) {
+      //cleanup after merge of reachable/unreachable values
+      for (Value value : unreachableValues.get(symbol)) {
+        reachableValues.remove(symbol, value);
+      }
+    }
+    //Merge states of values
+    for (Map.Entry<Value, State> valueStateEntry : executionState.stateOfValue.entrySet()) {
+      Value value = valueStateEntry.getKey();
+      State state = valueStateEntry.getValue();
+      State stateOfValue = getStateOfValue(value);
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Rename "stateOfValue" which hides the field declared at line 46.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)